platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/ReservedSignedState.java

```diff
@@ -0,0 +1,193 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.state.signed;
+
+import static com.swirlds.logging.LogMarker.EXCEPTION;
+
+import com.swirlds.common.AutoCloseableNonThrowing;
+import com.swirlds.common.exceptions.ReferenceCountException;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A wrapper around a signed state that holds a reservation. Until this wrapper is released/closed, the signed state
+ * contained within will not be destroyed.
+ * <p>
+ * This class is not thread safe. That is, it is not safe for one thread to access this object while another thread is
+ * asynchronously closing it. Each thread should hold its own reservation on a state if it needs to access a state.
+ */
+public class ReservedSignedState implements AutoCloseableNonThrowing {
+
+    private static final Logger logger = LogManager.getLogger(ReservedSignedState.class);
+
+    private final SignedState signedState;
+    private final String reason;
+    private final long reservationId = nextReservationId.getAndIncrement();
+    private boolean closed = false;
+
+    /**
+     * The next reservation id to use. It is ok that this is static, since we don't care which ID any particular
+     * reservation has, as long as that ID is unique.
+     */
+    private static final AtomicLong nextReservationId = new AtomicLong(0);
+
+    /**
+     * Create a wrapper around null.
+     */
+    public static @NonNull ReservedSignedState createNullReservation() {
+        return new ReservedSignedState();
+    }
+
+    /**
+     * Create a wrapper around null (for scenarios where we are storing a null signed state).
+     */
+    private ReservedSignedState() {
+        this.signedState = null;
+        this.reason = "";
+    }
+
+    /**
+     * Create a new reserved signed state.
+     *
+     * @param signedState the signed state to reserve
+     * @param reason      a short description of why this SignedState is being reserved. Each location where a
+     *                    SignedState is reserved should attempt to use a unique reason, as this makes debugging
+     *                    reservation bugs easier.
+     */
+    ReservedSignedState(@NonNull final SignedState signedState, @NonNull final String reason) {
+        this.signedState = Objects.requireNonNull(signedState);
+        this.reason = Objects.requireNonNull(reason);
+
+        // It is safe to "leak this" here.
```

Comment:
I'm not sure this is safe as the comment says. This class is not currently `final`. But even if it is, I don't think there is any guarantee that line 77 is executed executed prior to this line. It would be very easy to simply add a method that takes in the reason and reservation id instead of `this` for this call only.

Reply:
I've made this class final, was an oversight. Good suggestion, I no longer leak `this` at all.